### PR TITLE
Add message filters.

### PR DESCRIPTION
**Pull request by suneettipirneni**
Message filters are middleware that gets run whenever a message is sent to decide whether it should be deleted or not. A common example of this is deleting messages in a counting channel where a number is not sequential.

### CODE REVIEW

**Comment by rob-3** [comment]
Good feature overall. I think this fits better inside of the new `registerInteractionListener.ts` module in #18 which is responsible for managing incoming interactions and the state associated with `registerUI`. Under the design we're trying, I would pass in `registerMessageFilters` as a parameter to `run()` to remove the coupling to `Client` and keep SRP happy.

Also, I feel like you don't need to dynamically import message filters. They make the most sense to me as free functions in the file with the command that wants to implement them. This also will let us typecheck them at compile time before the signatures get erased.

**Comment by suneettipirneni** [comment]
> Good feature overall. I think this fits better inside of the new `registerInteractionListener.ts` module in #18 which is responsible for managing incoming interactions and the state associated with `registerUI`. Under the design we're trying, I would pass in `registerMessageFilters` as a parameter to `run()` to remove the coupling to `Client` and keep SRP happy.

That's fine

> 
> Also, I feel like you don't need to dynamically import message filters. They make the most sense to me as free functions in the file with the command that wants to implement them. This also will let us typecheck them at compile time before the signatures get erased.

I'd prefer to offer both options

**Comment by rob-3** [comment]
> I'd prefer to offer both options

What is the use case for dynamically importing? I feel like we just lose the type safety for nothing. The reason I feel the dynamically imported Commands are useful is so we don't have to keep an array of every command updated somewhere, but I feel like that would be a nonissue for the filters.

**Comment by suneettipirneni** [comment]
> > I'd prefer to offer both options
> 
> 
> 
> What is the use case for dynamically importing? I feel like we just lose the type safety for nothing. The reason I feel the dynamically imported Commands are useful is so we don't have to keep an array of every command updated somewhere, but I feel like that would be a nonissue for the filters.

Ok then I'll remove it.

**Comment by rob-3** [comment]
If you had something specific in mind, I'm not strongly against it or anything. For the uses I can think of though, dynamic imports seem like more trouble than they're worth.

**Comment by suneettipirneni** [comment]
> If you had something specific in mind, I'm not strongly against it or anything. For the uses I can think of though, dynamic imports seem like more trouble than they're worth.

I don't I just pretty much added it to mirror commands,  it commands are a different use case

**Comment by suneettipirneni** [comment]
Is this gtg?

**Comment by rob-3** [comment]
Almost. I will move some of the code once the `registerUI` branch is finalized and merged, but no reason this can't go in now. Only thing I would suggest is a slight change in API by making `runMessageFilters(message: Message, messageFilters: MessageFilter[])` into `runMessageFilter(message: Message, filter: MessageFilter)`.

File: src/messageFilters.ts [hunk 1]
```diff
@@ -0,0 +1,18 @@
+import { Message } from 'discord.js';
+
+export type MessageFilter = (message: Message) => boolean;
+
+export async function runMessageFilters(message: Message, messageFilters: MessageFilter[]): Promise<void> {
```
Comment (rob-3) [inline src/messageFilters.ts:5]: I would suggest just taking a single filter here and removing the looping logic.

File: src/Client.ts [hunk 2]
```diff
@@ -125,20 +127,25 @@ export default class Client extends discord.Client {
...
-      generatePermissionData(pushedCommands, this.commands.array());
+      generatePermissionData(pushedCommands, [...this.commands.values()]);
 
     // Apply Permissions (per-guild-only)
     await guild.commands.permissions.set({ fullPermissions });
   }
 
+  registerMessageFilters(filters: MessageFilter[]): void {
+    this.messageFilters.push(...filters);
+    this.on('messageCreate', async (message) => await runMessageFilters(message, this.messageFilters));
```
Comment (rob-3) [inline src/Client.ts:138]: You can use `this.messageFilters.forEach()` here to loop over all of the filters.

**Review by rob-3 — CHANGES_REQUESTED** [review]

File: src/messageFilters.ts (hunk 1 above)
Comment (suneettipirneni) [reply src/messageFilters.ts:5]: Why?

File: src/Client.ts (hunk 2 above)
Comment (suneettipirneni) [reply src/Client.ts:138]: The loop needs to be run sequentially, if a forEach() is used it does't run in sequence.

File: src/messageFilters.ts (hunk 1 above)
Comment (rob-3) [reply src/messageFilters.ts:5]: It pulls the looping responsibility outside of the function which I think is slightly better. The function doesn't really need to know whether there's multiple filters or not. I guess the justification imo is similar to why you wouldn't take `messages: Message[]`. Whether or not there's multiple doesn't really matter for the functionality (other than maybe this sequential execution thing below I need to read about).

File: src/Client.ts (hunk 2 above)
Comment (rob-3) [reply src/Client.ts:138]: Hmm so is your concern that the next iteration of the loop will start while the deletion is still happening?

Comment (suneettipirneni) [reply src/Client.ts:138]: Yes it causes a race condition, hence the for...of loop

Comment (rob-3) [reply src/Client.ts:138]: Ok I didn't realize this. I think it is good to merge then.

**Pull request merged**